MIP/mip-34/README.md

```diff
@@ -0,0 +1,261 @@
+The condition for slashing may be met by several quiteria, and not all slashing conditions may be used:
+- equivocate (send a different vote to different validators or users)
+- vote :white_check_mark: for an invalid block 
+- vote :x: for a valid block
+
+### Main challenges
+
+To achieve crypto-economically secured fast finality, we need to solve the following problems:
+
+1. design a _staking_ mechanism for the validators to stake assets, distribute rewards and manage slashing
+1. _define and verify_ the threshold (e.g. 2/3 of validators attest :white_check_mark:) for L2-finality
+1. _communicate_ the L2-finality status.
+
+In addition the following may require separate discussion, as it is a different procedure (postconfirmations are handled in smart contracts on L1, whereas L2-confirmations are handled off-chain)
+
+4. _define and verify_ the threshold (e.g. 2/3 of validators attest :white_check_mark:) for postconfirmation.
+
+
+### Components
+
+#### Staking (addresses 1. and 2.)
+
+The staking mechanism is implemented in a contract on L1.
+This contract provides the following functionalities:
+
+- join: a new validator can join the set of validators by staking some assets
+- exit: a validator exits and  get their stakes back
+- vote: receive a vote or a set of votes, verify the integrity of the votes (signatures) and the minimum threshold (e.g. 2/3)
+
+#### Generate an L2-finality certificate (addreses 3.)
+
+To ensure that the L2-finality status is made available to third-parties, we may publish our _proof_  (2/3 of attestations :white_check_mark:) to a data availability layer and get a _certificate_ that the proof is available.  
+This DA layer should offer a reliable _mempool_ for example as described [in this paper](https://arxiv.org/pdf/2105.11827).
+
+#### Handle postconfirmations (addresses 4.)
+
+The L1 contract will verify the L2-finality certificate. If the certificate is correct the block (or sequence of blocks) are _postconfirmed_. This requires handling who should send the certificate to the L1 contract, and how to verify the certificate.
+
+
+## Reference Implementation
+
+To simplify we assume that each validator stakes the same amount.
+The set of validators is in charge of validating sequenced batches and producing blocks that also commit to the state root of the sequenced batch.
+
+There may be different protocols for the postconfirmation and the L2-confirmation. Here we focus only on the L1 contract (i.e. postconfirmation).
+
+#### Version A: Leader-dependent blocks
+
+A leader validator is elected for a certain interval. The leader proposes the next transition (block-range $B_r'$):  $B_r \xrightarrow{\ txs \ } B_r'$. The leader can do so by sending a digest of $txs$ (Merkle root) and a digest of $B_r'$ (Merkle root hash of $B_r'$), or a _change set_. The leader commits on L1 to $B_r'$. Every validator checks the validity of $B_r'$ and prepares a vote message (:white_check_mark: or :x:).
```

Comment:
The leader election I think needs to be specified or otherwise acknowledged as a module in this system.

Reply:
It would be nice to be able to elect this leader on-chain--in the L1 contract.

Reply:
Also, we have talked about the possibility of a noble agent, i.e., Movement Labs, always being able to initiate this confirmation and other forms of volunteerism. 

The thing I worry about here are peer attacks. If a leader is elected, it's easier to target said leader than target everyone.

To avoid this, I think conceptually, we end up wanting to keep the volunteer set open. But, this essentially brings us back to [MD-4](https://github.com/movementlabsxyz/MIP/pulls).

The way I see it, in terms of consensus, the primary benefit of electing a leader is being able punish them for not doing their job. Accordingly, I think a stronger argument for leader penalties needs to be made here.